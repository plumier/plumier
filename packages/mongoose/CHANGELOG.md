# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.1.3](https://github.com/plumier/plumier/compare/v1.1.2...v1.1.3) (2023-03-10)

**Note:** Version bump only for package @plumier/mongoose





## [1.1.2](https://github.com/plumier/plumier/compare/v1.1.1...v1.1.2) (2023-01-20)

**Note:** Version bump only for package @plumier/mongoose





## [1.1.1](https://github.com/plumier/plumier/compare/v1.1.0...v1.1.1) (2023-01-03)

**Note:** Version bump only for package @plumier/mongoose





# [1.1.0](https://github.com/plumier/plumier/compare/v1.0.6...v1.1.0) (2022-12-19)

**Note:** Version bump only for package @plumier/mongoose





## [1.0.6](https://github.com/plumier/plumier/compare/v1.0.5...v1.0.6) (2021-10-04)

**Note:** Version bump only for package @plumier/mongoose





## [1.0.5](https://github.com/plumier/plumier/compare/v1.0.4...v1.0.5) (2021-07-25)

**Note:** Version bump only for package @plumier/mongoose





## [1.0.4](https://github.com/plumier/plumier/compare/v1.0.3...v1.0.4) (2021-07-05)

**Note:** Version bump only for package @plumier/mongoose





## [1.0.3](https://github.com/plumier/plumier/compare/v1.0.2...v1.0.3) (2021-06-28)


### Reverts

* Revert relation property readonly/writeonly on first class entity ([#966](https://github.com/plumier/plumier/issues/966)) ([bac445e](https://github.com/plumier/plumier/commit/bac445e8acee4c72e2c4096f74b6341549d80373))





## [1.0.2](https://github.com/plumier/plumier/compare/v1.0.0...v1.0.2) (2021-06-06)


### Bug Fixes

* Fix array relation should not accessible (read/write) on first class entity ([#956](https://github.com/plumier/plumier/issues/956)) ([58058a5](https://github.com/plumier/plumier/commit/58058a54861447d04cedfd585d60687eb3d4e1d4))
* Fix response authorization error when no ID provided on response ([#948](https://github.com/plumier/plumier/issues/948)) ([2b5429e](https://github.com/plumier/plumier/commit/2b5429ef30f9cfb3843fb07c5af271dd3223b14c))





## 1.0.1 (2021-05-18)

**Note:** Version bump only for package @plumier/mongoose
