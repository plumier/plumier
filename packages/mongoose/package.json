{
  "name": "@plumier/mongoose",
  "version": "1.0.0-beta.11",
  "description": "Mongoose with decorator based configuration",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "reflection",
    "introspection"
  ],
  "scripts": {
    "compile": "tsc -p tsconfig.build.json"
  },
  "author": "Ketut Sandiarsa",
  "license": "MIT",
  "dependencies": {
    "@plumier/core": "1.0.0-beta.11",
    "@types/mongoose": "^5.5.32",
    "chalk": "^3.0.0",
    "mongoose": "^5.7.13"
  },
  "bugs": {
    "url": "https://github.com/plumier/plumier/issues"
  },
  "repository": {
    "url": "https://github.com/plumier/plumier"
  },
  "homepage": "https://plumierjs.com",
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "9d69b5621b7e2e777818e5e36c13ecba0c8d8204"
}
