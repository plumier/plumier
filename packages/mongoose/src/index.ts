import {
    Class,
    DefaultFacility,
    domain,
    findFilesRecursive,
    isCustomClass,
    PlumierApplication,
    ValidatorDecorator,
    val
} from "@plumier/core"
import Chalk from "chalk"
import Mongoose, { Model } from "mongoose"
import { dirname, isAbsolute, join } from "path"
import {
    ClassReflection,
    decorateClass,
    decorateProperty,
    mergeDecorator,
    PropertyReflection,
    reflect,
    Reflection,
} from "tinspector"
import { VisitorInvocation, Result } from "typedconverter"
import { safeToString } from 'typedconverter/lib/converter';

/* ------------------------------------------------------------------------------- */
/* ------------------------------------ TYPES ------------------------------------ */
/* ------------------------------------------------------------------------------- */

export type Constructor<T> = new (...args: any[]) => T
export type SchemaRegistry = { [key: string]: Mongoose.Schema }
export interface MongooseFacilityOption {
    model?: string | Class | Class[]
    uri: string,
    schemaGenerator?: SchemaGenerator
}
export interface SubSchema {
    type: typeof Mongoose.Schema.Types.ObjectId,
    ref: string
}
export type SchemaGenerator = (definition: any, meta: ClassReflection) => Mongoose.Schema
interface AnalysisResult {
    type: "warning" | "error",
    message: string
}
interface DomainAnalysis {
    domain: ClassReflection
    analysis: AnalysisResult[]
}
interface MongooseCollectionDecorator {
    type: "MongooseCollectionDecorator",
    alias?: string
}

const GlobalMongooseSchema: SchemaRegistry = {}
const ArrayHasNoTypeInfo = `MONG1000: Array property {0}.{1} require @array(<Type>) decorator to be able to generated into mongoose schema`
const NoClassFound = `MONG1001: No class decorated with @collection() found`
const CanNotValidateNonProperty = `MONG1002: @val.unique() only can be applied on property`
const ModelNotDecoratedWithCollection = `MONG1003: {0} not decorated with @collection()`

/* ------------------------------------------------------------------------------- */
/* ------------------------------- SCHEMA GENERATOR ------------------------------ */
/* ------------------------------------------------------------------------------- */

function loadModels(opt: Class[]) {
    return opt.map(x => reflect(x))
}

function getType(prop: PropertyReflection, registry: SchemaRegistry): Function | Function[] | SubSchema | SubSchema[] {
    if (isCustomClass(prop.type)) {
        const schema = { type: Mongoose.Schema.Types.ObjectId, ref: "" }
        return Array.isArray(prop.type) ? [{ ...schema, ref: getName(prop.type[0]) }]
            : { ...schema, ref: getName(prop.type) }
    }
    else return prop.type
}

function generateModel(model: ClassReflection, registry: SchemaRegistry, generator?: SchemaGenerator) {
    const definition = model.properties
        .reduce((a, b) => {
            a[b.name] = getType(b, registry)
            return a
        }, {} as any)
    registry[getName(model)] = !!generator ? generator(definition, model) : new Mongoose.Schema(definition)
}

function generateSchema(opt: Class[], registry: SchemaRegistry, generator?: SchemaGenerator) {
    loadModels(opt).forEach(x => generateModel(x, registry, generator))
}

/* ------------------------------------------------------------------------------- */
/* --------------------------------- ANALYZER ------------------------------------ */
/* ------------------------------------------------------------------------------- */

function noArrayTypeInfoTest(domain: ClassReflection): AnalysisResult[] {
    return domain.properties
        .map(x => (x.type === Array) ?
            <AnalysisResult>{ message: ArrayHasNoTypeInfo.format(domain.name, x.name), type: "error" } : undefined)
        .filter((x): x is AnalysisResult => Boolean(x))
}

function analyze(domains: ClassReflection[]) {
    const tests = [noArrayTypeInfoTest]
    return domains.map(x => (<DomainAnalysis>{
        domain: x,
        analysis: tests.map(test => test(x)).flatten()
    }))
}

function printAnalysis(analysis: DomainAnalysis[]) {
    console.log()
    console.log(Chalk.bold("Model Analysis Report"))
    if (!analysis.map(x => x.domain).some(x =>
        x.decorators.some((y: MongooseCollectionDecorator) => y.type === "MongooseCollectionDecorator"))) {
        console.log(NoClassFound)
    }
    else {
        const namePad = Math.max(...analysis.map(x => x.domain.name.length))
        analysis.forEach((x, i) => {
            const num = (i + 1).toString().padStart(analysis.length.toString().length)
            const color = x.analysis.some(x => x.type === "error") ? Chalk.red : (x: string) => x
            console.log(color(`${num}. ${x.domain.name.padEnd(namePad)} -> ${getName(x.domain)}`))
            x.analysis.forEach(y => {
                console.log(Chalk.red(`  - ${y.type} ${y.message}`))
            })
        })
    }
}

/* ------------------------------------------------------------------------------- */
/* --------------------------------- HELPERS ------------------------------------- */
/* ------------------------------------------------------------------------------- */

async function isUnique(value: string, target: Class | undefined, field: string) {
    if (!target) throw new Error(CanNotValidateNonProperty)
    const Model = model(target)
    const condition: { [key: string]: object } = {}
    //case insensitive comparison
    condition[field] = { $regex: value, $options: "i" }
    const result = await Model.findOne(condition)
    if (!!result) return `${value} already exists`
}

/* ------------------------------------------------------------------------------- */
/* ------------------------------- MAIN FUNCTIONS -------------------------------- */
/* ------------------------------------------------------------------------------- */

declare module "typedconverter" {
    namespace val {
        function unique(): (target: any, name: string, index?: any) => void
    }
}

val.unique = () => val.custom(async (value, info) => {
    if(info.ctx.method.toLocaleLowerCase() === "post")
        return isUnique(value, info.parent && info.parent.type, info.name)
})

export function reflectPath(path: string | Class | Class[]): Reflection[] {
    if (Array.isArray(path))
        return path.map(x => reflect(x))
    else if (typeof path === "string")
        return findFilesRecursive(path)
            .map(x => reflect(x))
            .map(x => x.members)
            .flatten()
    else
        return [reflect(path)]
}

export function collection(alias?: string) {
    return mergeDecorator(decorateClass(<MongooseCollectionDecorator>{ type: "MongooseCollectionDecorator", alias }),
        reflect.parameterProperties());
}

export function getName(opt: ClassReflection | Class) {
    const meta = typeof opt === "function" ? reflect(opt) : opt
    const decorator = meta.decorators.find((x: MongooseCollectionDecorator): x is MongooseCollectionDecorator => x.type === "MongooseCollectionDecorator")
    return decorator && decorator.alias || meta.name
}

function relationToObjectIdVisitor(i: VisitorInvocation): Result {
    const id = safeToString(i.value)
    if (Mongoose.Types.ObjectId.isValid(id) && isCustomClass(i.type))
        return Result.create(Mongoose.Types.ObjectId(id))
    else
        return i.proceed()
}

export class MongooseFacility extends DefaultFacility {
    option: MongooseFacilityOption
    constructor(opts: MongooseFacilityOption) {
        super()
        const model = opts.model || "./model"
        const domain = typeof model === "string" ? isAbsolute(model) ?
            model! : join(dirname(module.parent!.filename), model) : model
        this.option = { ...opts, model: domain }
    }

    async initialize(app: Readonly<PlumierApplication>) {
        //generate schemas
        const collections = reflectPath(this.option.model!)
            .filter((x): x is ClassReflection => x.kind === "Class")
            .filter(x => x.decorators.some((x: MongooseCollectionDecorator) => x.type == "MongooseCollectionDecorator"))
        if (app.config.mode === "debug") {
            const analysis = analyze(collections)
            printAnalysis(analysis)
        }
        generateSchema(collections.map(x => x.type), GlobalMongooseSchema, this.option.schemaGenerator)
        //register custom converter
        app.set({ typeConverterVisitors: [relationToObjectIdVisitor] })
        Mongoose.set("useUnifiedTopology", true)
        await Mongoose.connect(this.option.uri, { useNewUrlParser: true  })
    }
}

export function model<T>(type: Constructor<T>) {

    function traversePropertyType(meta: ClassReflection): Class[] {
        const properties = meta.properties
            .filter(x => isCustomClass(x.type))
            .map((x) => <Class>(Array.isArray(x.type) ? x.type[0] : x.type))
        return properties.length > 0 ?
            properties.concat(properties.map(x => reflect(x))
                .map(x => traversePropertyType(x))
                .flatten()) : []
    }

    class ModelProxyHandler<T> implements ProxyHandler<Mongoose.Model<T & Mongoose.Document>> {
        private isLoaded = false
        modelName: string;
        metaData: ClassReflection

        constructor(domain: Constructor<T>) {
            const meta = reflect(domain)
            this.modelName = getName(meta)
            this.metaData = meta
        }

        private getModelByName(name: string): Mongoose.Model<T & Mongoose.Document> {
            if (!Mongoose.connection.models[name])
                return Mongoose.model(name, GlobalMongooseSchema[name])
            else
                return Mongoose.model(name)
        }

        private getModel(): Mongoose.Model<T & Mongoose.Document> {
            if (!this.isLoaded) {
                const properties = traversePropertyType(this.metaData)
                const unique = Array.from(new Set(properties))
                unique.forEach(x => this.getModelByName(getName(x)))
                this.isLoaded = true
            }
            return this.getModelByName(this.modelName)
        }

        get(target: Mongoose.Model<T & Mongoose.Document>, p: PropertyKey, receiver: any): any {
            if (GlobalMongooseSchema[this.modelName]) {
                const Model = this.getModel();
                return (Model as any)[p]
            }
            else {
                return p === "toString" ? () => "[Function]" : (target as any)[p]
            }
        }

        construct?(target: Mongoose.Model<T & Mongoose.Document>, argArray: any, newTarget?: any): object {
            try {
                const Model = this.getModel();
                return new Model(...argArray)
            } catch (e) {
                throw new Error(ModelNotDecoratedWithCollection.format(this.modelName))
            }
        }
    }
    return new Proxy(Mongoose.Model as Mongoose.Model<T & Mongoose.Document>, new ModelProxyHandler<T>(type))
}
