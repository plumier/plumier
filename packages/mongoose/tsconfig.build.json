{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "declaration": true,
    "rootDir": "src",
    "outDir": "lib",
    "sourceMap": false,
    "skipLibCheck": true
  },
  "exclude": [
    "test",
    "lib"
  ]
}