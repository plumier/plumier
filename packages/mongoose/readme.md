# Plumier Mongoose Helper

[![Build Status](https://github.com/plumier/plumier/workflows/ubuntu/badge.svg)](https://github.com/plumier/plumier/actions?query=workflow%3Aubuntu)
[![Build status](https://github.com/plumier/plumier/workflows/windows/badge.svg)](https://github.com/plumier/plumier/actions?query=workflow%3Awindows)
[![Coverage Status](https://coveralls.io/repos/github/plumier/plumier/badge.svg?branch=master)](https://coveralls.io/github/plumier/plumier?branch=master)
[![Greenkeeper badge](https://badges.greenkeeper.io/plumier/plumier.svg)](https://greenkeeper.io/)
[![lerna](https://img.shields.io/badge/maintained%20with-lerna-cc00ff.svg)](https://lernajs.io/)
![npm (canary)](https://img.shields.io/npm/v/plumier/canary)
![npm (latest)](https://img.shields.io/npm/v/plumier/latest)



## Description
Plumier Mongoose Helper help you easily map your domain model and create Mongoose model using it. Helper automatically generate schema definition based on your domain model metadata.

- [x] Pure POJO, used conditional type to increase type inference on nested documents.
- [x] Same schema configuration used by mongoose
- [x] Inheritance
- [x] Field properties or parameter properties declaration
- [x] Model analysis

```typescript
import { model, collection } from "@plumier/mongoose"

// base class, all derived class will inherit the behavior
@collection({ timestamp: true, toJSON: { virtuals: true } })
class Domain {
    @collection.property()
    id:string
    @collection.property({ default:false })
    deleted:boolean
    @collection.property()
    createdAt:Date
    @collection.property()
    updatedAt:Date
}

@collection()
class User extends Domain{
    constructor(
        public name:string,
        @collection.property({ unique:true })
        public email:string,
        public dateOfBirth: Date
    ) { super() }
}

// create mongoose model
const UserModel = model(User)

@collection()
class UserActivity extends Domain {
    constructor(
        @collection.ref(User)
        public user: User,
        @collection.property({ default: () => new Date() })
        public date: Date,
        public browser:string,
        public os:string,
        @reflect.types([Number])
        public latLong: number[]
    ){ super() }
}

// create mongoose model
const UserActivityModel = model(UserActivity)
```

## Domain Model Declaration 
Plumier Mongoose Helper uses tinspector to extract type metadata on runtime. Currently there are two domain models declaration supported

### Using Property Field 
```typescript
@collection()
class Dummy {
    @collection.property()
    stringProp: string

    @collection.property()
    numberProp: number

    @collection.property()
    booleanProp: boolean

    @collection.property()
    dateProp: Date
}
```

This is the common model declaration when you are familiar with Nest.js or other TypeScript framework. This declaration required `strictPropertyInitialization` disabled on `tsconfig.json` file. Note that the `@collection.property()` is required when there are no decorator applied on the property. 

### Using TypeScript Parameter Properties

```typescript 
import { collection } from "@plumier/mongoose"

@collection()
class Dummy {
    constructor(
        public stringProp: string,
        public numberProp: number,
        public booleanProp: boolean,
        public dateProp: Date
    ) { }
}
```

This declaration good when `tsconfig.json` uses `strict: true` because we unable to use field properties. Using this declaration reduce the need of using `@collection.properties()` on all properties. 

## Features 

### Basic Schema Generation 

```typescript
import { model, collection } from "@plumier/mongoose"

@collection()
class Dummy {
    constructor(
        public stringProp: string,
        public numberProp: number,
        public booleanProp: boolean,
        public dateProp: Date
    ) { }
}
// create Mongoose model by using below
const DummyModel = model(Dummy)
// example usage
const result = await DummyModel.findById(<id>)
// model can be called multiple time to create other model instance
// for example from inside controller or in other place
const SecondDummyModel = model(Dummy)
```

### Advanced Data Type 
Array type required extra decorator information. Use `@reflect.type([<type>])` decorator to inform generator about extra type information.

```typescript
import { model, collection } from "@plumier/mongoose"

@collection()
class Child {
    constructor(
        public name:string
    ){}
}

@collection()
class Dummy {
    constructor(
        // primitive array
        @reflect.type([String])
        public arrayOfPrimitive: string[],

        // nested type 
        public child: Child,

        // nested array of model 
        @reflect.type([Child])
        public children: Child[],
    ) { }
}

const DummyModel = model(Dummy)
```

### Nested Document With Ref (Populate)
```typescript
import { model, collection } from "@plumier/mongoose"

@collection()
class Child {
    constructor(
        public name:string
    ){}
}

@collection()
class Dummy {
    constructor(
        // nested type 
        @collection.ref(Child)
        public child: Child,

        // nested array of model 
        @collection.ref([Child])
        public children: Child[],
    ) { }
}

const ChildModel = model(Child)
const DummyModel = model(Dummy)

const dummy = await DummyModel.findById("").populate("child")
// child inferred as Document, so its possible to call Document related props/methods
dummy.child.id
```

### Configure Properties 
Extra Mongoose schema configuration can be passed to each decorator 

```typescript
import { model, collection } from "@plumier/mongoose"

@collection({ timestamps: true, toJSON: { virtuals: true } })
class Dummy {
    constructor(
        @collection.property({ uppercase:true })
        public stringProp: string,
        @collection.property({ unique:true })
        public email: string,
        @collection.property({ default:() => new Date() })
        public dateProp: Date
    ) { }
}

const DummyModel = model(Dummy)
```

### Inheritance 
Inheritance work naturally, all child document will inherit parent configuration properly. 

```typescript
import { model, collection } from "@plumier/mongoose"

@collection({ timestamp: true })
class DomainBase {
    @collection.property({ default:false })
    deleted:boolean
    @collection.property()
    createdAt:Date
    @collection.property()
    updatedAt:Date
}

@collection()
class Dummy extends DomainBase{
    constructor(
        public stringProp: string,
        public numberProp: number,
        public booleanProp: boolean,
        public dateProp: Date
    ) { super() }
}

const DummyModel = model(Dummy)
```

Using configuration above, all class inherited from `DomainBase` will have `deleted` property with default value `false` and properties `createdAt` and `updatedAt` which automatically populated as timestamps.

### Custom Model Name 

```typescript
import { model, collection } from "@plumier/mongoose"

@collection()
class Dummy {
    constructor(
        public stringProp: string,
        public numberProp: number,
        public booleanProp: boolean,
        public dateProp: Date
    ) { }
}
// use the second parameter to change the model name
const DummyModel = model(Dummy, "Empty")
// calling next model will not require passing name
const SecondDummyModel = model(Dummy)
```

### Schema Generation Hook

```typescript
import { model, collection } from "@plumier/mongoose"
import mongoose from "mongoose"

@collection()
class Dummy {
    constructor(
        public stringProp: string,
        public numberProp: number,
        public booleanProp: boolean,
        public dateProp: Date
    ) { }
}
// pass function on second parameter to hook schema generation
const DummyModel = model(Dummy, schema => {
    schema.pre("save", next => {
        // do something
        next()
    })
})
// calling next model will not require passing the hook
const SecondDummyModel = model(Dummy)
```

### Model Analysis 
Mongoose helper has built-in model analysis contains information about model, configuration and their appropriate MongoDB collection, use it like below


```typescript
import { model, collection, printAnalysis, getAnalysis } from "@plumier/mongoose"

@collection()
class Dummy {
    constructor(
        public stringProp: string,
        public numberProp: number,
        public booleanProp: boolean,
        public dateProp: Date
    ) { }
}
const DummyModel = model(Dummy)

// somewhere when your code starting
printAnalysis(getAnalysis())
```

## Dockify
`Dockify<T>` is an advanced TypeScript type, it converts all Property of `T` inherit from `Object` into mongoose `Document`. For example: 

```typescript 
class Child {
    constructor(
        public name:string
    ){}
}

class Parent {
    constructor(
        public child:Child
    ){}
}

let parent:Dockify<Parent>
// child property converted into `Child & mongoose.Document` 
// thus its possible access document properties/method like below
parent.child._id
parent.child.save()
```

`Dockify<T>` provide syntax sugar to access ref (populate) properties, while keep entity definition POJO (clean from Mongoose specific types). 

> **CAVEAT**: Dockify will treat all properties with custom type as Document, thus for non ref (populate) property will keep inferred as `Document`. 


