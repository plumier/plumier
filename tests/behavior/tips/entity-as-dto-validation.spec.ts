import { collection } from "@plumier/mongoose"
import mongoose from "mongoose"
import Plumier, { authorize, route, val, WebApiFacility, CustomValidatorFunction as Val } from "plumier"
import supertest = require("supertest")

@collection()
export class User {
    constructor(
        @val.required()
        public name: string,
        @val.required()
        public password: string,
        public dateOfBirth: Date,
        public active: boolean
    ) { }
}

const validatePassword: Val = (value, meta) => {
    if (value.password !== value.confirmPassword)
        return val.result("confirmPassword", "Password doesn't match")
}

export class UsersController {
    @authorize.public()
    @route.post("")
    save(@val.custom(validatePassword) data: User) {
        return data
    }
}

const app = new Plumier()
    .set(new WebApiFacility({ controller: __filename }))
    .set({ mode: "production" })

afterAll(async () => await mongoose.disconnect())

it("Should run correctly", async () => {
    const koa = await app.initialize()
    await supertest(koa.callback())
        .post("/users")
        .send({ name: "Lorem", password: "abcd", confirmPassword: "abcd" })
        .expect(200)
})

it("Should return 422 when try to set role", async () => {
    const koa = await app.initialize()
    const { body } = await supertest(koa.callback())
        .post("/users")
        .send({ name: "Lorem", password: "abcd", confirmPassword: "123" })
        .expect(422)
    expect(body).toMatchSnapshot()
})