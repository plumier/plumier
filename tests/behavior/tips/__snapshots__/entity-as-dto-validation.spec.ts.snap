// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should return 422 when try to set role 1`] = `
Object {
  "message": Array [
    Object {
      "messages": Array [
        "Password doesn't match",
      ],
      "path": Array [
        "data",
        "confirmPassword",
      ],
    },
  ],
  "status": 422,
}
`;
